### PR TITLE
p2p: Reduce inv traffic during IBD

### PR DESCRIPTION
**Pull request by maflcko**
Tx-inv messages are ignored during IBD, so it would be nice if we told peers to not send them in the first place. Do that by sending two `feefilter` messages: One when the connection is made (and the node is in IBD), and another one when the node leaves IBD.

### CODE REVIEW

**Review by luke-jr — APPROVED** [review]
utACK

**Comment by hebasto** [comment]
Concept ACK.

File: src/net_processing.cpp [hunk 1]
```diff
@@ -4394,9 +4394,20 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
             !pto->HasPermission(PF_FORCERELAY)) {
             CAmount currentFilter = m_mempool.GetMinFee(gArgs.GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000).GetFeePerK();
             int64_t timeNow = GetTimeMicros();
+            static FeeFilterRounder filterRounder{CFeeRate{DEFAULT_MIN_RELAY_TX_FEE}};
```
Comment (sipa) [inline src/net_processing.cpp:4397]: Nit: as you're touching all places where this variable is used, convert it to current naming style?

Also: can this be `const`?

File: src/net_processing.cpp [hunk 2]
```diff
@@ -4394,9 +4394,20 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
...
+            if (m_chainman.ActiveChainstate().IsInitialBlockDownload()) {
+                // Reveived tx-inv messages are discarded when the active
```
Comment (sipa) [inline src/net_processing.cpp:4399]: Typo: reveived

File: test/functional/rpc_net.py [hunk 3]
```diff
@@ -44,7 +44,6 @@ def assert_net_servicesnames(servicesflag, servicenames):
 
 class NetTest(BitcoinTestFramework):
     def set_test_params(self):
-        self.setup_clean_chain = True
```
Comment (sipa) [inline test/functional/rpc_net.py:47]: Unrelated change?

**Comment by sipa** [comment]
utACK fa78c686ca5bcafa153d2f3813b77449fa31bb89. Nice idea.

File: src/validation.h [hunk 4]
```diff
@@ -799,7 +799,8 @@ class ChainstateManager
     std::vector<CChainState*> GetAll();
 
     //! The most-work chain.
-    CChain& ActiveChain() const;
+    CChainState& ActiveChainstate() const;
+    CChain& ActiveChain() const { return ActiveChainstate().m_chain; }
```
Comment (hebasto) [inline src/validation.h:803]: If `ActiveChain()` definition is placed in header, why `ActiveChainstate()` definition is not placed here too?

**Comment by practicalswift** [comment]
Concept ACK: neat idea! :)

File: src/net_processing.cpp [hunk 5]
```diff
@@ -4394,9 +4394,20 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
             !pto->HasPermission(PF_FORCERELAY)) {
             CAmount currentFilter = m_mempool.GetMinFee(gArgs.GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000).GetFeePerK();
             int64_t timeNow = GetTimeMicros();
+            static FeeFilterRounder filterRounder{CFeeRate{DEFAULT_MIN_RELAY_TX_FEE}};
+            if (m_chainman.ActiveChainstate().IsInitialBlockDownload()) {
```
Comment (hebasto) [inline src/net_processing.cpp:4398]: The code that actually rejects tx-inv messages https://github.com/bitcoin/bitcoin/blob/6762a627ecb89ba8d4ed81a049a5d802e6dd75c2/src/net_processing.cpp#L2584-L2586

contains `fImporting` and `fReindex`.

Should block importing and reindexing be considered as well here?

File: src/net_processing.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/net_processing.cpp:4397]: `filterRounder` has a fast random context, which is not const. Calling `round` on it will change the state.

File: src/validation.h (hunk 4 above)
Comment (MarcoFalke) [reply src/validation.h:803]: To minimize the diff and aid review

File: src/net_processing.cpp (hunk 5 above)
Comment (MarcoFalke) [reply src/net_processing.cpp:4398]: `IsInitialBlockDownload` *does* consider when blocks are imported

File: src/test/policy_fee_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+BOOST_AUTO_TEST_CASE(FeeRounder)
+{
+    FeeFilterRounder fee_rounder{CFeeRate{1000}};
+
+    // check that 1000 rounds to 974 or 1071
+    std::set<CAmount> results;
+    while (results.size() < 2) {
+        results.emplace(fee_rounder.round(1000));
+    }
+    BOOST_CHECK_EQUAL(*results.begin(), 974);
```
Comment (rajarshimaitra) [inline src/test/policy_fee_tests.cpp:23]: What is the purpose of fee rounder and why it will return 974 and 1071?

**Review by rajarshimaitra — COMMENTED** [review]
Concept ACK. Neat Idea. Compiled without issue, all tests passing. 
Was just wondering `rpc_net` is not really checking this situation, probably an unrelated change? Also, would it be possible to try having a functional test for this behavior? 

Following is just a question for my own understanding.

File: src/test/policy_fee_tests.cpp (hunk 6 above)
Comment (MarcoFalke) [reply src/test/policy_fee_tests.cpp:23]: There are privacy concerns with deanonymizing a node by the fact that it is broadcasting identifying information about its mempool min fee. To help ameliorate this concern, the implementation quantizes the filter value broadcast with a small amount of randomness, in addition, the messages are broadcast to different peers at individually randomly distributed times.

Source: https://github.com/bitcoin/bips/blob/master/bip-0133.mediawiki#considerations

**Comment by MarcoFalke** [comment]
Thanks for the review so far! The latest force push comes with those changes:

* Fix typo and global variable naming
* Document why one of the functional tests needs to change
* Add a new refactoring cleanup commit, removing three global symbols and replacing them with one local

**Comment by laanwj** [comment]
Concept ACK

> Why we don't pretend it's a blocks-only connection at the beginning? I guess maybe because it's unknown before handshake.

One problem with that is that it's impossible to switch a blocks-only connection to a full connection. So it'd require dropping and reconnecting *all* connections after IBD. I'm not sure that's a good idea.

> I'd feel uncomfortable using bip37 for this. bloomfilters have been hidden behind a run time flag. It seems fragile to assume that full node implementations in the future are more likely to support bip37 message types than the feefilter message type.

However I don't think @naumenkogs is suggesting to use BIP37 messages here, but the `fRelay` bit in the version message.

**Comment by MarcoFalke** [comment]
Jup, fRelay can't be changed while keeping the connection because subsequent version messages are ignored. Recycling the connection seems overly complex and potentially fragile.

File: src/net_processing.cpp (hunk 1 above)
Comment (sipa) [reply src/net_processing.cpp]: Uh, that means it needs a mutex (or very clearly stated assumptions that no two PeerLogicValidation objects can exist that could be accessed from separate threads).

Comment (MarcoFalke) [reply src/net_processing.cpp]: This seems unrelated to the changes here. Filed under a new issue: #19254

Comment (sipa) [reply src/net_processing.cpp]: Cool, thanks.

**Review by glozow — COMMENTED** [review]
Giant Concept ACK. I think doing this via `feefilter` is pretty elegant, since we can't change connection type, can't send a second version message with a new `fRelay`, and can't switch from `blocksonly` mode.
There's apparently a way to do this via `fRelay` by putting `fRelay=False` in the version message and then sending a `filterclear`later ([#8709](https://github.com/bitcoin/bitcoin/pull/8709)). In my opinion, nodes not serving bloomfilters should just disconnect on `filterclear`.
An idea that I'm not super confident in: what about adding a P2P message to toggle `fRelay` on and off? (I think this is mentioned in gmaxwell's [comment](https://github.com/bitcoin/bitcoin/pull/8709#issuecomment-246957287) on #8709) and seems naumenkogs mentioned it too. I'm not sure what does/doesn't warrant a P2P protocol change.

**Comment by MarcoFalke** [comment]
A new message type wouldn't be understood by old nodes, making it less effective than the `feefilter` approach. Also, it wouldn't be more flexible, because `feefilter` can already be used to effectively turn off and on transaction relay at any time.

**Comment by gmaxwell** [comment]
Using the fee filter for this is clever and I think a good idea.

Though disconnecting all peers at the end of IBD would not be bad: Currently IBD staller eviction can make you end up with a bias towards fast and nearby peers-- this effect would be worse but the eviction is far from aggressive enough and that has gone without being fixed in part because fixing it would further bias peer selection. As a result IBD times can easily vary by many times based on how lucky you got with your peer selection.

At some point bitcoind should probably switch to continually kicking off the least contributing peer during IBD once per minute regardless of stalls then rescue its topology when it gets out of IBD by dropping all (or all but 1/2 peers).  If that were done, these connections could be fRelay=0.

**Comment by naumenkogs** [comment]
utACK fa8a66cf7e255884cb5cf061f43f42a7371e7ff4

**Review by hebasto — APPROVED** [review]
ACK fa8a66cf7e255884cb5cf061f43f42a7371e7ff4, tested on Linux Mint 19.3 (x86_64):

```
$ ./test/functional/rpc_net.py
2020-06-12T07:26:00.331000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_nf7udywg
2020-06-12T07:26:00.801000Z TestFramework (INFO): Get out of IBD for the minfeefilter test
2020-06-12T07:26:00.805000Z TestFramework (INFO): Connect nodes both way
2020-06-12T07:26:01.008000Z TestFramework (INFO): Connect nodes both way
2020-06-12T07:26:01.453000Z TestFramework (INFO): Stopping nodes
2020-06-12T07:26:01.605000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_nf7udywg on exit
2020-06-12T07:26:01.605000Z TestFramework (INFO): Tests successful
```

**Comment by ajtowns** [comment]
utACK fa8a66cf7e255884cb5cf061f43f42a7371e7ff4

File: src/net_processing.cpp [hunk 7]
```diff
@@ -2581,7 +2581,7 @@ bool ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRec
                     LogPrint(BCLog::NET, "transaction (%s) inv sent in violation of protocol, disconnecting peer=%d\n", inv.hash.ToString(), pfrom.GetId());
                     pfrom.fDisconnect = true;
                     return true;
-                } else if (!fAlreadyHave && !fImporting && !fReindex && !::ChainstateActive().IsInitialBlockDownload()) {
+                } else if (!fAlreadyHave && !chainman.ActiveChainstate().IsInitialBlockDownload()) {
```
Comment (jonatack) [inline src/net_processing.cpp:2584]: fa3b6974 I think the commit message would be clearer if it stated that `!fImporting && !fReindex` can be removed here because they are checked in `IsInitialBlockDownload` which returns true if `fImporting || fReindex`

**Review by jonatack — COMMENTED** [review]
ACK fa8a66cf7e255884 a functional test in `p2p_feefilter.py` for the changes in fa8a66cf would be a good idea

**Comment by MarcoFalke** [comment]
If someone wants to write a functional test for this, I am happy to review. Otherwise, I will write one myself after merge.

File: src/net_processing.cpp (hunk 7 above)
Comment (MarcoFalke) [reply src/net_processing.cpp:2584]: I keep my commit messages as brief as possible and try to only include meta information that can not be derived from the raw source code. My goal is to encourage reviewers to give closer looks instead of trusting the commit message.

**Comment by jamesob** [comment]
ACK fa525e4d1cfda8c1924d2c69f43bd7ae3b98fb72 ([`jamesob/ackr/19204.1.MarcoFalke.p2p_reduce_inv_traffic_d`](https://github.com/jamesob/bitcoin/tree/ackr/19204.1.MarcoFalke.p2p_reduce_inv_traffic_d))

Haven't tested. Tried to build locally, but got a linker failure (which sounds
like is not unique to this branch): 
```
/usr/bin/ld: bitcoin_wallet-bitcoin-wallet.o:(.data.rel.ro+0x8): undefined reference to `InitError(bilingual_str const&)'
```

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-4.19.0-9-amd64-x86_64-with-glibc2.28

Configured with ./configure LDFLAGS=-L/home/james/src/bitcoin/db4/lib/ CPPFLAGS=-I/home/james/src/bitcoin/db4/include/ CXXFLAGS=-std=c++11 --enable-wallet --enable-debug --with-daemon

Compiled with /usr/bin/ccache /usr/bin/g++ -std=c++11 -mavx -mavx2 -std=c++11 -fno-extended-identifiers -O0 -g3 -ftrapv  -fstack-reuse=none -Wstack-protector -fstack-protector-all -msse4 -msha -msse4.1 -msse4.2  i

Compiler version: g++ (Debian 8.3.0-6) 8.3.0
```

</p></details>

<details><summary>Show signature data</summary>
<p>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK fa525e4d1cfda8c1924d2c69f43bd7ae3b98fb72 ([`jamesob/ackr/19204.1.MarcoFalke.p2p_reduce_inv_traffic_d`](https://github.com/jamesob/bitcoin/tree/ackr/19204.1.MarcoFalke.p2p_reduce_inv_traffic_d))

Haven't tested. Tried to build locally, but got a linker failure (which sounds
like is not unique to this branch): 
```
/usr/bin/ld: bitcoin_wallet-bitcoin-wallet.o:(.data.rel.ro+0x8): undefined reference to `InitError(bilingual_str const&)'
```

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-4.19.0-9-amd64-x86_64-with-glibc2.28

Configured with ./configure LDFLAGS=-L/home/james/src/bitcoin/db4/lib/ CPPFLAGS=-I/home/james/src/bitcoin/db4/include/ CXXFLAGS=-std=c++11 --enable-wallet --enable-debug --with-daemon

Compiled with /usr/bin/ccache /usr/bin/g++ -std=c++11 -mavx -mavx2 -std=c++11 -fno-extended-identifiers -O0 -g3 -ftrapv  -fstack-reuse=none -Wstack-protector -fstack-protector-all -msse4 -msha -msse4.1 -msse4.2  i

Compiler version: g++ (Debian 8.3.0-6) 8.3.0
```

</p></details>
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEpm2+LMqPsWj2rAwWyrywS1A+5mgFAl7s1u0ACgkQyrywS1A+
5mha8A/8DH0r6iY6lZbK9pkU7JHO+z+AGrvHG+oNq6e7GiXQqNQdgAZifhh9kfiv
mH+z1JZPX4pTTiC9945BtfWLI1apjZsgOslFlhTFCwP0STbvGxcBYyGO7k/HqUHg
X03Au2iSRH02avbeyd4oCxbBIy2wZJZSkK7Aezhtq9m6C8rgNrFtDHi9JYAGDys/
fMj4p/z+3DaIldeFivYWcT6fXyy6AtNXSJlWlxJFk2divc46V/5haR0s0KtyyQyt
UNAAOf830rVEb6yl14HFov3HRWB2V6HHf8NTkP9vRo3GYuGqhyPqrEa4+MVIcnaA
qAVQEzMfCuj6XTOyh3Yx8nterGG4L6+ak9TfagdjgoAme6lenfQN5IbLeEMUXsZL
RxbT+bAwgAbZ1OP+J+XAt74jbAsnIw27pew+KpnnpaDwQAHiZLkFy7zRGA1LSYkI
jm9NyNx7I0YPwSR2/K1q6OpsoVIubgMrP3Sm4boqOnmGSc3ZZEURdJDwxoFFnwc9
pbGX4yRIxvUT31CvWnWIeF2zo9RN29Y4mzwufUMyLqoj3NqD6GZ6H/nuMkRbDT4d
jr17KilECwi6IBMRthZn6a7el3fRQe4leGBEjwpbIh78q5SoU445VXLSJIS28Ts2
1Lrwngh3GkHq8UDMrIY/hpzsO9HOnP1Xvcpjn3RHdSFBcO03AGE=
=hJ4k
-----END PGP SIGNATURE-----

```

</p></details>

**Comment by MarcoFalke** [comment]
Jup, the gcc bug is unrelated and tracked in #19309

**Review by glozow — COMMENTED** [review]
ACK https://github.com/bitcoin/bitcoin/commit/fa525e4d1cfda8c1924d2c69f43bd7ae3b98fb72
I tested it functionally with [this](https://github.com/gzhao408/bitcoin/commit/495d6a86bff86e41cb15836994894502bb8515b2), confirms that the minfeefilter is set before/after IBD.

> If someone wants to write a functional test for this, I am happy to review. Otherwise, I will write one myself after merge.

I'm not sure if this counts as a complete functional test 😅   this is kind of a draft that just checks minfeefilter (not any txrelay behavior) but I'd be down to do this.

**Comment by MarcoFalke** [comment]
For anyone who reviewed before the rebase, it should be trivial to re-ACK with `git range-diff bitcoin-core/master A B`

**Comment by naumenkogs** [comment]
utACK fa525e4

**Review by jonatack — COMMENTED** [review]
re-ACK fa525e4 checked diff `git range-diff 19612ca fa8a66c fa525e4`, re-reviewed, ran tests, ran a custom p2p IBD behavior test at https://github.com/jonatack/bitcoin/commit/9321e0f223ea87d21f6fa42b61bcc8d40a0943de.

@gzhao408 in that link I adapted your test, feel free to use any or all of it for a PR.

**Review by hebasto — APPROVED** [review]
re-ACK fa525e4d1cfda8c1924d2c69f43bd7ae3b98fb72, only rebased since the [previous](https://github.com/bitcoin/bitcoin/pull/19204#pullrequestreview-429519667) review (verified with `git range-diff`).

**Comment by rebroad** [comment]
Wouldn't the filterclear message be better to use for when tx invs are desired? I could be wrong, but I think this would (or should) request tx invs even if the connection was made with relay=0 (in the version message).

**Comment by MarcoFalke** [comment]
Have you seen https://github.com/bitcoin/bitcoin/pull/19204#issuecomment-642009022 and #19260 ?

**Comment by rebroad** [comment]
> Concept ACK
> 
> > Why we don't pretend it's a blocks-only connection at the beginning? I guess maybe because it's unknown before handshake.
> 
> One problem with that is that it's impossible to switch a blocks-only connection to a full connection. So it'd require dropping and reconnecting _all_ connections after IBD. I'm not sure that's a good idea.
> 
> > I'd feel uncomfortable using bip37 for this. bloomfilters have been hidden behind a run time flag. It seems fragile to assume that full node implementations in the future are more likely to support bip37 message types than the feefilter message type.
> 
> However I don't think @naumenkogs is suggesting to use BIP37 messages here, but the `fRelay` bit in the version message.

Actually, I don't think it's impossible - the filterclear message can be sent to change from a blocks-only connection to one that allows TXs.

**Comment by rebroad** [comment]
> Have you seen [#19204 (comment)](https://github.com/bitcoin/bitcoin/pull/19204#issuecomment-642009022) and #19260 ?

yes, I have, but the bloomfilters hidden behind a runtime flag surely did not disable the filterclear message, did they? They should not have - as this was by design the way to allow TXs to be sent after IBD had finished.
